src/natcap/invest/hra.py

```diff
@@ -1,3253 +1,2424 @@
+            ``target_srs_wkt``.
+        target_pixel_size (tuple): A 2-element tuple of the (x, y) pixel size
+            of the target raster.  Elements are in units of the target SRS.
+        target_pixel_type (int): The GDAL GDT_* type of the target raster.
+        target_srs_wkt (string): The SRS of the target raster, in Well-Known
+            Text format.
+        target_nodata (float): If provided, the nodata value of the target
+            raster.
+        fill_value=None (number): If provided, the value that the target raster
+            should be filled with.
+
+    Returns:
+        ``None``
+    """
+    bbox_minx, bbox_miny, bbox_maxx, bbox_maxy = target_bounding_box
+
+    driver = gdal.GetDriverByName('GTiff')
+    n_bands = 1
+    n_cols = int(numpy.ceil(
+        abs((bbox_maxx - bbox_minx) / target_pixel_size[0])))
+    n_rows = int(numpy.ceil(
+        abs((bbox_maxy - bbox_miny) / target_pixel_size[1])))
+
+    raster = driver.Create(
+        target_raster_path, n_cols, n_rows, n_bands, target_pixel_type,
+        options=['TILED=YES', 'BIGTIFF=YES', 'COMPRESS=DEFLATE',
+                 'BLOCKXSIZE=256', 'BLOCKYSIZE=256'])
+    raster.SetProjection(target_srs_wkt)
+
+    # Set the transform based on the upper left corner and given pixel
+    # dimensions.  Bounding box is in format [minx, miny, maxx, maxy]
+    if target_pixel_size[0] < 0:
+        x_source = bbox_maxx
+    else:
+        x_source = bbox_minx
+    if target_pixel_size[1] < 0:
+        y_source = bbox_maxy
+    else:
+        y_source = bbox_miny
+    raster_transform = [
+        x_source, target_pixel_size[0], 0.0,
+        y_source, 0.0, target_pixel_size[1]]
+    raster.SetGeoTransform(raster_transform)
+
+    # Fill the band if requested.
+    band = raster.GetRasterBand(1)
+    if fill_value is not None:
+        band.Fill(fill_value)
+
+    # Set the nodata value.
+    if target_nodata is not None:
+        band.SetNoDataValue(float(target_nodata))
+
+    band = None
+    raster = None
+
+
+def _simplify(source_vector_path, tolerance, target_vector_path,
+              preserve_columns=None):
+    """Simplify a geometry to a given tolerance.
+
+    This function uses the GEOS SimplifyPreserveTopology function under the
+    hood.  For docs, see GEOSTopologyPreserveSimplify() at
+    https://libgeos.org/doxygen/geos__c_8h.html
+
+    Args:
+        source_vector_path (string): The path to a source vector to simplify.
+        tolerance (number): The numerical tolerance to simplify by.
+        target_vector_path (string): Where the simplified geometry should be
+            stored on disk.
+        preserve_columns=None (iterable or None): If provided, this is an
+            iterable of string column names (case-insensitive) that should be
+            carried over from the source vector to the target vector.  If
+            ``None``, no columns will be carried over.
+
+    Returns:
+        ``None``.
+    """
+    if preserve_columns is None:
```

Comment:
I haven't observed it taking a long time, but it's always possible that it could if the user provides a complicated geometry.  Adding a logging statement in f1074330b